File: connector/src/main/java/org/geysermc/connector/network/translators/world/WorldManager.java
/*
 * Copyright (c) 2019-2021 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.connector.network.translators.world;

import com.github.steveice10.mc.protocol.data.game.chunk.Chunk;
import com.github.steveice10.mc.protocol.data.game.entity.metadata.Position;
import com.github.steveice10.mc.protocol.data.game.entity.player.GameMode;
import com.github.steveice10.mc.protocol.data.game.setting.Difficulty;
import com.nukkitx.math.vector.Vector3i;
import com.nukkitx.nbt.NbtMap;
import org.geysermc.connector.network.session.GeyserSession;
import org.geysermc.connector.utils.GameRule;

/**
 * Class that manages or retrieves various information
 * from the world. Everything in this class should be
 * safe to return null or an empty value in the event
 * that chunk caching or anything of the sort is disabled
 * on the standalone version of Geyser.
 */
public abstract class WorldManager {

    /**
     * Gets the Java block state at the specified location
     *
     * @param session the session
     * @param position the position
     * @return the block state at the specified location
     */
    public int getBlockAt(GeyserSession session, Position position) {
        return this.getBlockAt(session, position.getX(), position.getY(), position.getZ());
    }

    /**
     * Gets the Java block state at the specified location
     *
     * @param session the session
     * @param vector the position
     * @return the block state at the specified location
     */
    public int getBlockAt(GeyserSession session, Vector3i vector) {
        return this.getBlockAt(session, vector.getX(), vector.getY(), vector.getZ());
    }

    /**
     * Gets the Java block state at the specified location
     *
     * @param session the session
     * @param x the x coordinate to get the block at
     * @param y the y coordinate to get the block at
     * @param z the z coordinate to get the block at
     * @return the block state at the specified location
     */
    public abstract int getBlockAt(GeyserSession session, int x, int y, int z);

    /**
     * Gets all block states in the specified chunk section.
     *
     * @param session the session
     * @param x the chunk's X coordinate
     * @param y the chunk's Y coordinate
     * @param z the chunk's Z coordinate
     * @param section the chunk section to store the block data in
     */
    public abstract void getBlocksInSection(GeyserSession session, int x, int y, int z, Chunk section);

    /**
     * Checks whether or not this world manager requires a separate chunk cache/has access to more block data than the chunk cache.
     * <p>
     * Some world managers (e.g. Spigot) can provide access to block data outside of the chunk cache, and even with chunk caching disabled. This
     * method provides a means to check if this manager has this capability.
     *
     * @return whether or not this world manager has access to more block data than the chunk cache
     */
    public abstract boolean hasOwnChunkCache();

    /**
     * Gets the Java biome data for the specified chunk.
     *
     * @param session the session of the player
     * @param x the chunk's X coordinate
     * @param z the chunk's Z coordinate
     * @return the biome data for the specified region with a length of 1024.
     */
    public abstract int[] getBiomeDataAt(GeyserSession session, int x, int z);

    /**
     * Sigh. <br>
     *
     * So, on Java Edition, the lectern is an inventory. Java opens it and gets the contents of the book there.
     * On Bedrock, the lectern contents are part of the block entity tag. Therefore, Bedrock expects to have the contents
     * of the lectern ready and present in the world. If the contents are not there, it takes at least two clicks for the
     * lectern to update the tag and then present itself. <br>
     *
     * We solve this problem by querying all loaded lecterns, where possible, and sending their information in a block entity
     * tag.
     *
     * @param session the session of the player
     * @param x the x coordinate of the lectern
     * @param y the y coordinate of the lectern
     * @param z the z coordinate of the lectern
     * @param isChunkLoad if this is called during a chunk load or not. Changes behavior in certain instances.
     * @return the Bedrock lectern block entity tag. This may not be the exact block entity tag - for example, Spigot's
     * block handled must be done on the server thread, so we send the tag manually there.
     */
    public abstract NbtMap getLecternDataAt(GeyserSession session, int x, int y, int z, boolean isChunkLoad);

    /**
     * @return whether we should expect lectern data to update, or if we have to fall back on a workaround.
     */
    public abstract boolean shouldExpectLecternHandled();

    /**
     * Updates a gamerule value on the Java server
     *
     * @param session The session of the user that requested the change
     * @param name The gamerule to change
     * @param value The new value for the gamerule
     */
    public abstract void setGameRule(GeyserSession session, String name, Object value);

    /**
     * Gets a gamerule value as a boolean
     *
     * @param session The session of the user that requested the value
     * @param gameRule The gamerule to fetch the value of
     * @return The boolean representation of the value
     */
    public abstract Boolean getGameRuleBool(GeyserSession session, GameRule gameRule);

    /**
     * Get a gamerule value as an integer
     *
     * @param session The session of the user that requested the value
     * @param gameRule The gamerule to fetch the value of
     * @return The integer representation of the value
     */
    public abstract int getGameRuleInt(GeyserSession session, GameRule gameRule);

    /**
     * Change the game mode of the given session
     *
     * @param session The session of the player to change the game mode of
     * @param gameMode The game mode to change the player to
     */
    public abstract void setPlayerGameMode(GeyserSession session, GameMode gameMode);

    /**
     * Change the difficulty of the Java server
     *
     * @param session The session of the user that requested the change
     * @param difficulty The difficulty to change to
     */
    public abstract void setDifficulty(GeyserSession session, Difficulty difficulty);

    /**
     * Checks if the given session's player has a permission
     *
     * @param session The session of the player to check the permission of
     * @param permission The permission node to check
     * @return True if the player has the requested permission, false if not
     */
    public abstract boolean hasPermission(GeyserSession session, String permission);
}

File: core/src/main/java/org/geysermc/geyser/level/WorldManager.cpp
/*
 * Copyright (c) 2019-2022 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.geyser.level;

#include "io.netty.buffer.ByteBuf"
#include "io.netty.buffer.Unpooled"
#include "it.unimi.dsi.fastutil.ints.Int2ObjectMap"
#include "it.unimi.dsi.fastutil.ints.Int2ObjectMaps"
#include "org.checkerframework.checker.nullness.qual.NonNull"
#include "org.checkerframework.checker.nullness.qual.Nullable"
#include "org.cloudburstmc.math.vector.Vector3i"
#include "org.geysermc.erosion.util.BlockPositionIterator"
#include "org.geysermc.geyser.level.block.type.BlockState"
#include "org.geysermc.geyser.session.GeyserSession"
#include "org.geysermc.mcprotocollib.protocol.data.game.entity.player.GameMode"
#include "org.geysermc.mcprotocollib.protocol.data.game.item.component.DataComponent"
#include "org.geysermc.mcprotocollib.protocol.data.game.item.component.DataComponentType"
#include "org.geysermc.mcprotocollib.protocol.data.game.item.component.DataComponentTypes"
#include "org.geysermc.mcprotocollib.protocol.data.game.item.component.DataComponents"
#include "org.geysermc.mcprotocollib.protocol.data.game.setting.Difficulty"

#include "java.util.HashMap"
#include "java.util.List"
#include "java.util.Locale"
#include "java.util.Map"
#include "java.util.concurrent.CompletableFuture"
#include "java.util.function.Consumer"
#include "java.util.function.Function"


public abstract class WorldManager {


    public final BlockState blockAt(GeyserSession session, Vector3i vector) {
        return this.blockAt(session, vector.getX(), vector.getY(), vector.getZ());
    }


    public BlockState blockAt(GeyserSession session, int x, int y, int z) {
        return BlockState.of(this.getBlockAt(session, x, y, z));
    }


    public final int getBlockAt(GeyserSession session, Vector3i vector) {
        return this.getBlockAt(session, vector.getX(), vector.getY(), vector.getZ());
    }


    public abstract int getBlockAt(GeyserSession session, int x, int y, int z);

    public final CompletableFuture<Integer> getBlockAtAsync(GeyserSession session, Vector3i vector) {
        return this.getBlockAtAsync(session, vector.getX(), vector.getY(), vector.getZ());
    }

    public CompletableFuture<Integer> getBlockAtAsync(GeyserSession session, int x, int y, int z) {
        return CompletableFuture.completedFuture(this.getBlockAt(session, x, y, z));
    }

    public int[] getBlocksAt(GeyserSession session, BlockPositionIterator iter) {
        int[] blocks = new int[iter.getMaxIterations()];
        for (; iter.hasNext(); iter.next()) {
            int networkId = this.getBlockAt(session, iter.getX(), iter.getY(), iter.getZ());
            blocks[iter.getIteration()] = networkId;
        }
        return blocks;
    }


    public abstract bool hasOwnChunkCache();


    public void setGameRule(GeyserSession session, std::string name, Object value) {
        session.sendCommandPacket("gamerule " + name + " " + value);
    }


    public abstract bool getGameRuleBool(GeyserSession session, GameRule gameRule);


    public abstract int getGameRuleInt(GeyserSession session, GameRule gameRule);


    public abstract GameMode getDefaultGameMode(GeyserSession session);


    public void setDefaultGameMode(GeyserSession session, GameMode gameMode) {
        session.sendCommandPacket("defaultgamemode " + gameMode.name().toLowerCase(Locale.ROOT));
    }


    public void setDifficulty(GeyserSession session, Difficulty difficulty) {
        session.sendCommandPacket("difficulty " + difficulty.name().toLowerCase(Locale.ROOT));
    }


    public std::string [] getBiomeIdentifiers(bool withTags) {
        return null;
    }


    public void getDecoratedPotData(GeyserSession session, Vector3i pos, Consumer<List<std::string>> apply) {
    }

    protected static final Function<Int2ObjectMap<byte[]>, DataComponents> RAW_TRANSFORMER = map -> {
        try {
            Map<DataComponentType<?>, DataComponent<?, ?>> components = new HashMap<>();
            Int2ObjectMaps.fastForEach(map, entry -> {
                DataComponentType<?> type = DataComponentTypes.from(entry.getIntKey());
                ByteBuf buf = Unpooled.wrappedBuffer(entry.getValue());
                DataComponent<?, ?> value = type.readDataComponent(buf);
                components.put(type, value);
            });
            return new DataComponents(components);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    };
}
